code/game/objects/random/random.dm

```diff
@@ -1531,27 +1535,45 @@
 	concealable = TRUE
 
 /obj/random/weapon_and_ammo/post_spawn(var/obj/item/gun/projectile/spawned)
-	if(!istype(spawned, /obj/item/gun/projectile))
+	if(istype(spawned, /obj/item/gun/energy))
 		return
-	if(spawned.magazine_type)
-		var/obj/item/ammo_magazine/am = spawned.magazine_type
-		new am(spawned.loc)
-		new am(spawned.loc)
-	else if(istype(spawned, /obj/item/gun/projectile/shotgun) && spawned.caliber == "shotgun")
-		if(istype(spawned.loc, /obj/item/storage/box))
-			spawned.loc.icon_state = "largebox"
-		var/obj/item/storage/box/b = new /obj/item/storage/box(spawned.loc)
-		for(var/i = 0; i < 8; i++)
-			new spawned.ammo_type(b)
-	else if(spawned.ammo_type)
-		var/list/provided_ammo = list()
-		for(var/i = 0; i < (spawned.max_shells * 2); i++)
-			provided_ammo += new spawned.ammo_type(spawned.loc)
-		if(provided_ammo.len)
-			new /obj/item/ammo_pile(spawned.loc, provided_ammo)
-
-	if(istype(spawned, /obj/item/gun/projectile/musket))
-		new /obj/item/reagent_containers/powder_horn(spawned.loc)
+
+	else if(istype(spawned, /obj/item/gun/projectile))
+		if(spawned.magazine_type)
+			var/obj/item/ammo_magazine/am = spawned.magazine_type
+			new am(spawned.loc)
+			new am(spawned.loc)
+		else if(istype(spawned, /obj/item/gun/projectile/shotgun) && spawned.caliber == "shotgun")
+			if(istype(spawned.loc, /obj/item/storage/box))
+				spawned.loc.icon_state = "largebox"
+			var/obj/item/storage/box/b = new /obj/item/storage/box(spawned.loc)
+			for(var/i = 0; i < 8; i++)
+				new spawned.ammo_type(b)
+		else if(spawned.ammo_type)
+			var/list/provided_ammo = list()
+			for(var/i = 0; i < (spawned.max_shells * 2); i++)
+				provided_ammo += new spawned.ammo_type(spawned.loc)
+			if(provided_ammo.len)
+				new /obj/item/ammo_pile(spawned.loc, provided_ammo)
+
+		if(istype(spawned, /obj/item/gun/projectile/musket))
+			new /obj/item/reagent_containers/powder_horn(spawned.loc)
+
+	else if(istype(spawned, /obj/item/gun/launcher))
```

Comment:
All of these istypes should instead be a proc on the gun that spawns stuff there for the future

Reply:
Good idea